bundle/edu.gemini.catalog/src/main/scala/edu/gemini/catalog/image/StoredImagesCache.scala

```diff
@@ -0,0 +1,161 @@
+    case x if x._2 === i => (Instant.now, x._2)
+    case x               => x
+  })
+
+  /**
+    * Return images sorted by access time
+    */
+  def sortedByAccess: List[ImageInFile] = entries.sortBy(_._1).map(_._2).reverse
+
+  // We need a special order for finding the best match for images that
+  // takes into account the image distance and size
+  val imageInFileOrder: Order[(Angle, ImageInFile)] =
+    Order.orderBy(i => (i._1, i._2.query.size))
+
+  /**
+    * Find the image in the cache closest to the requested query
+    */
+  def closestImage(query: ImageSearchQuery): Option[ImageInFile] = {
+    val distances = for {
+        (_, e)   <- entries
+        if e.query.catalog === query.catalog
+        distance = query.coordinates.angularDistance(e.query.coordinates)
+        if distance <= ImageSearchQuery.maxDistance
+      } yield (distance, e)
+
+    distances.minimum(imageInFileOrder).map(_._2)
+  }
+
+  /**
+    * Return the closest image in the cache containing containing the query
+    */
+  def inside(query: ImageSearchQuery): Option[ImageInFile] = {
+    val distances = for {
+        (_, e) <- entries
+        if e.query.catalog === query.catalog
+        if e.contains(query.coordinates)
+      } yield (query.coordinates.angularDistance(e.query.coordinates), e)
+
+    distances.minimum(imageInFileOrder).map(_._2)
+  }
+
+}
+
+object StoredImages {
+  val zero = StoredImages(Nil)
+
+  /** @group Typeclass Instances */
+  implicit val equals: Equal[StoredImages] = Equal.equalA[StoredImages]
+}
+
+/**
+  * In memory cache of images on disk
+  */
+object StoredImagesCache {
+  private val cacheRef = TaskRef.newTaskRef[StoredImages](StoredImages.zero).unsafePerformSync
+
+  def add(i: ImageInFile): Task[StoredImages] = cacheRef.mod(_ + i) *> cacheRef.get
+
+  def addAt(instant: Instant, i: ImageInFile): Task[StoredImages] = cacheRef.mod(_ + (instant, i)) *> cacheRef.get
+
+  def markAsUsed(i: ImageInFile): Task[StoredImages] = cacheRef.mod(_.touch(i)) *> cacheRef.get
+
+  def remove(i: ImageInFile): Task[StoredImages] = cacheRef.mod(_ - i) *> cacheRef.get
+
+  def get: Task[StoredImages] = cacheRef.get
+
+  def clean: Task[StoredImages] = cacheRef.mod(_ => StoredImages.zero) *> cacheRef.get
+
+  /**
+    * Find if the search query is in the cache, returning the closest image if present
+    */
+  def find(query: ImageSearchQuery): Task[Option[ImageInFile]] =
+    cacheRef.get.map(_.inside(query))
+}
+
+/**
+  * Useful methods to update the cached files
+  */
+object ImageCacheOnDisk {
+
+  /**
+    * Method to prune the cache, limiting the space used
+    */
+  def pruneCache(maxSize: Information): Task[Unit] = Task.fork {
+    // Remove files from the in memory cache and delete from drive
+    def deleteOldFiles(files: List[ImageInFile]): Task[Unit] =
+      Task.gatherUnordered(files.map(StoredImagesCache.remove)) *> Task.delay(files.foreach(_.file.toFile.delete()))
```

Comment:
The tasks in this list aren't associated with an executor so they will run sequentially when you do this. But all they're doing is an in-memory update so maybe you just want `files.traverseU(StoredImagesCache.remove)`?